File: platform/workers/ingestion-premium/src/identity.ts
/** Contract-driven natural key, event_time, available_at, and persist run identity. */

import type { DatasetSpec } from "./catalog";
import { sha256HexFromString } from "./sha256.ts";

/** Persist/SCD2 run identity. crypto.randomUUID, not Math.random. */
export function newRunId(prefix: string): string {
  return `${prefix}-${crypto.randomUUID()}`;
}

const DATE_RE = /^\d{4}-\d{2}-\d{2}$/;

export function isYyyyMmDd(value: string): boolean {
  return DATE_RE.test(value);
}

function validDate(value: string): boolean {
  if (!isYyyyMmDd(value)) return false;
  const date = new Date(`${value}T00:00:00Z`);
  return !Number.isNaN(date.getTime()) && date.toISOString().slice(0, 10) === value;
}

const JST_OFFSET_MS = 9 * 60 * 60 * 1000;

export function toJstIso(d: Date): string {
  const ms = d.getTime() + JST_OFFSET_MS;
  const jst = new Date(ms);
  return jst.toISOString().replace(/\.(\d+)Z$/, "+09:00");
}

export function todayJst(): string {
  return toJstIso(new Date()).slice(0, 10);
}

export function daysAgoJst(n: number): string {
  const t = new Date(Date.now() - n * 24 * 60 * 60 * 1000);
  return toJstIso(t).slice(0, 10);
}

function jsonString(value: string): string {
  return JSON.stringify(value);
}

/** Python mirror: data_contracts.identity.canonical_json. */
export function stableJson(value: unknown): string {
  if (value === null) return "null";
  if (typeof value === "boolean" || typeof value === "number" || typeof value === "string") {
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) {
    return `[${value.map((item) => stableJson(item === undefined ? null : item)).join(",")}]`;
  }
  if (typeof value === "object") {
    const object = value as Record<string, unknown>;
    return `{${Object.keys(object).filter((key) => object[key] !== undefined).sort().map(
      (key) => `${jsonString(key)}:${stableJson(object[key])}`,
    ).join(",")}}`;
  }
  throw new Error(`not an interoperable JSON value: ${typeof value}`);
}

function aliasesFor(spec: DatasetSpec, field: string): string[] {
  return [field, ...(spec.field_aliases?.[field] ?? [])];
}

function pick(row: Record<string, unknown>, spec: DatasetSpec, field: string): unknown | null {
  for (const candidate of aliasesFor(spec, field)) {
    for (const key of [candidate, candidate.toLowerCase()]) {
      const value = row[key];
      if (value !== undefined && value !== null && value !== "") return value;
    }
  }
  return null;
}

export async function naturalKey(
  row: Record<string, unknown>,
  spec: DatasetSpec,
): Promise<string> {
  const picked: Record<string, unknown> = {};
  for (const field of spec.natural_key_fields) {
    const value = pick(row, spec, field);
    if (value === null || value === "") {
      return `hash:sha256:${await sha256HexFromString(stableJson(row))}`;
    }
    picked[field] = value;
  }
  return stableJson(picked);
}

export function sessionCloseJst(dateYyyyMmDd: string, session?: string): string {
  if (!validDate(dateYyyyMmDd)) {
    throw new Error(`expected YYYY-MM-DD, got ${JSON.stringify(dateYyyyMmDd)}`);
  }
  const close = session === "morning"
    ? "11:30:00"
    : dateYyyyMmDd < "2024-11-05" ? "15:00:00" : "15:30:00";
  return `${dateYyyyMmDd}T${close}+09:00`;
}

function dateStart(value: unknown): string | null {
  return typeof value === "string" && validDate(value)
    ? `${value}T00:00:00+09:00`
    : null;
}

function validClock(value: string): boolean {
  const match = /^(\d{2}):(\d{2})(?::(\d{2})(?:\.\d+)?)?$/.exec(value);
  if (!match) return false;
  return Number(match[1]) < 24 && Number(match[2]) < 60 && Number(match[3] ?? 0) < 60;
}

function timestampFromFields(
  row: Record<string, unknown>,
  spec: DatasetSpec,
  fields: string[],
): string | null {
  if (fields.length === 0) return null;
  const first = pick(row, spec, fields[0]);
  if (typeof first !== "string" || first.length === 0) return null;
  if (first.includes("T") || first.includes(" ")) return first;
  if (fields.length < 2 || !validDate(first)) return null;
  const second = pick(row, spec, fields[1]);
  if (typeof second !== "string" || !validClock(second)) return null;
  const clock = /^\d{2}:\d{2}$/.test(second) ? `${second}:00` : second;
  return `${first}T${clock}+09:00`;
}

export function pickEventTime(
  row: Record<string, unknown>,
  spec: DatasetSpec,
): string | null {
  const fields = spec.event_time_fields;
  if (fields.length === 0) return null;
  if (spec.event_time_policy === "session_close") {
    const day = pick(row, spec, fields[0]);
    return typeof day === "string" && validDate(day)
      ? sessionCloseJst(day, spec.session)
      : null;
  }
  if (spec.event_time_policy === "explicit_timestamp_field") {
    return timestampFromFields(row, spec, fields) ?? dateStart(pick(row, spec, fields[0]));
  }
  if (spec.event_time_policy === "observation_date") {
    return dateStart(pick(row, spec, fields[0]));
  }
  return null;
}

function nextCalendarStart(value: unknown): string | null {
  if (typeof value !== "string" || !validDate(value)) return null;
  const day = new Date(`${value}T00:00:00Z`);
  day.setUTCDate(day.getUTCDate() + 1);
  return `${day.toISOString().slice(0, 10)}T00:00:00+09:00`;
}

function knownLagAvailableAt(
  row: Record<string, unknown>,
  spec: DatasetSpec,
): string | null {
  const event = pickEventTime(row, spec);
  const lag = spec.known_publication_lag;
  if (!event || !lag) return null;
  const match = /^P(?:(\d+)D)?(?:T(?:(\d+)H)?)?$/.exec(lag);
  if (!match) return null;
  const date = new Date(event);
  if (Number.isNaN(date.getTime())) return null;
  date.setTime(date.getTime() + (Number(match[1] ?? 0) * 24 + Number(match[2] ?? 0)) * 3_600_000);
  // Preserve an explicit offset representation for SQLite lexical ordering.
  const shifted = new Date(date.getTime() + 9 * 3_600_000).toISOString().replace("Z", "+09:00");
  return shifted.replace(".000+", "+");
}

export function pickAvailableAt(
  row: Record<string, unknown>,
  spec: DatasetSpec,
  ingestedAt: string,
): string {
  const policy = spec.available_at_policy;
  if (policy === "session_close") {
    const day = pick(row, spec, spec.availability_field ?? "Date");
    if (typeof day === "string" && validDate(day)) {
      return sessionCloseJst(day, spec.session);
    }
  } else if (policy === "explicit_timestamp_field" && spec.availability_field) {
    const instant = timestampFromFields(row, spec, spec.availability_field.split("+"));
    if (instant) return instant;
  } else if (policy === "explicit_disclosure_date" && spec.availability_field) {
    const fields = spec.availability_field.split("+");
    const instant = timestampFromFields(row, spec, fields);
    if (instant) return instant;
    const conservativeDayEnd = nextCalendarStart(pick(row, spec, fields[0]));
    if (conservativeDayEnd) return conservativeDayEnd;
  } else if (policy === "known_publication_lag") {
    const lagged = knownLagAvailableAt(row, spec);
    if (lagged) return lagged;
  }
  return ingestedAt;
}

File: packages/data_plane/data_contracts/identity.py
"""Contract-driven, cross-runtime row identity and PIT timestamps.

The TypeScript mirror imports the same JSON contract.  ``canonical_json`` is
defined to match JavaScript JSON number rendering and UTF-16 object-key order,
which makes the SHA-256 fallback byte-identical in Python and Workers.

Finite numbers follow ECMAScript ``JSON.stringify`` (ordinary fractions,
negatives, ``-0``, the ``1e-6`` / ``1e21`` exponent thresholds, and integers
coerced through binary64).  Python-generated canonical bytes, payload strings,
and digests from before the fractional-number fix are not trustworthy when they
contain affected fractions such as ``0.45`` / ``0.11``; this module does not
rewrite stored rows.
"""

from __future__ import annotations

from datetime import datetime, timedelta, timezone
import hashlib
import json
import math
import re
from typing import Any, Mapping, Sequence

from .loader import DatasetContract, contract_for
from .source_capability import (
    SourceCapabilityContract,
    source_capability_contract_for,
)

JST = timezone(timedelta(hours=9))
_DATE_RE = re.compile(r"^\d{4}-\d{2}-\d{2}$")
_LAG_RE = re.compile(r"^P(?:(\d+)D)?(?:T(?:(\d+)H)?)?$")


def _utf16_sort_key(value: str) -> bytes:
    return value.encode("utf-16-be", errors="surrogatepass")


def _json_string(value: str) -> str:
    return json.dumps(value, ensure_ascii=False, separators=(",", ":"))


def _js_number(value: int | float) -> str:
    """Render a Python JSON number like ECMAScript ``JSON.stringify``.

    J-Quants JSON uses interoperable finite numbers.  Coercing integers through
    binary64 intentionally matches JavaScript's single ``number`` type.  Stripping
    leading zeros from the digit string must also move the decimal point;
    otherwise ``0.45`` is emitted as ``4.5``.
    """
    number = float(value)
    if not math.isfinite(number):
        return "null"
    if number == 0:
        return "0"
    negative = number < 0
    raw = repr(abs(number)).lower()
    if "e" in raw:
        mantissa, exp_text = raw.split("e", 1)
        exponent = int(exp_text)
    else:
        mantissa, exponent = raw, 0
    integer, _dot, fraction = mantissa.partition(".")
    digits = integer + fraction
    leading_zeros = len(digits) - len(digits.lstrip("0"))
    digits = digits.lstrip("0") or "0"
    decimal_pos = len(integer) + exponent - leading_zeros
    magnitude = abs(number)
    if 1e-6 <= magnitude < 1e21:
        if decimal_pos <= 0:
            rendered = "0." + ("0" * -decimal_pos) + digits
        elif decimal_pos >= len(digits):
            rendered = digits + ("0" * (decimal_pos - len(digits)))
        else:
            rendered = digits[:decimal_pos] + "." + digits[decimal_pos:]
        if "." in rendered:
            rendered = rendered.rstrip("0").rstrip(".")
    else:
        significant = digits.rstrip("0") or "0"
        sci_exp = decimal_pos - 1
        rendered = significant[0]
        if len(significant) > 1:
            rendered += "." + significant[1:]
        rendered += "e" + ("+" if sci_exp >= 0 else "") + str(sci_exp)
    return ("-" if negative else "") + rendered


def canonical_json(value: Any) -> str:
    """Serialize an interoperable JSON value to deterministic UTF-8 text."""
    if value is None:
        return "null"
    if value is True:
        return "true"
    if value is False:
        return "false"
    if isinstance(value, (int, float)):
        return _js_number(value)
    if isinstance(value, str):
        return _json_string(value)
    if isinstance(value, (list, tuple)):
        return "[" + ",".join(canonical_json(v) for v in value) + "]"
    if isinstance(value, Mapping):
        items: list[str] = []
        for key in sorted((str(k) for k in value), key=_utf16_sort_key):
            items.append(_json_string(key) + ":" + canonical_json(value[key]))
        return "{" + ",".join(items) + "}"
    raise TypeError(f"not an interoperable JSON value: {type(value).__name__}")


def sha256_fallback(row: Mapping[str, Any]) -> str:
    digest = hashlib.sha256(canonical_json(row).encode("utf-8")).hexdigest()
    return f"hash:sha256:{digest}"


def _pick(
    row: Mapping[str, Any], contract: DatasetContract, field: str
) -> Any | None:
    for candidate in contract.aliases_for(field):
        for key in (candidate, candidate.lower()):
            value = row.get(key)
            if value is not None and value != "":
                return value
    return None


def natural_key(row: Mapping[str, Any], dataset_id: str) -> str:
    """Return the contract-selected natural key or SHA-256 row fallback.

    Composite keys are all-or-nothing.  A missing discriminator uses the row
    hash instead of a partial key that could collapse distinct observations.
    """
    contract = contract_for(dataset_id)
    picked: dict[str, Any] = {}
    for field in contract.natural_key_fields:
        value = _pick(row, contract, field)
        if value is None or value == "":
            return sha256_fallback(row)
        picked[field] = value
    return canonical_json(picked)


def session_close_jst(date_yyyy_mm_dd: str, *, session: str | None = None) -> str:
    _parse_date(date_yyyy_mm_dd)
    if session == "morning":
        close = "11:30:00"
    else:
        close = "15:00:00" if date_yyyy_mm_dd < "2024-11-05" else "15:30:00"
    return f"{date_yyyy_mm_dd}T{close}+09:00"


def _parse_date(value: str) -> datetime:
    if not _DATE_RE.fullmatch(value):
        raise ValueError(f"expected YYYY-MM-DD, got {value!r}")
    try:
        return datetime.strptime(value, "%Y-%m-%d").replace(tzinfo=JST)
    except ValueError as exc:
        raise ValueError(f"expected YYYY-MM-DD, got {value!r}") from exc


def _date_start(value: Any) -> str | None:
    if not isinstance(value, str) or not _DATE_RE.fullmatch(value):
        return None
    try:
        _parse_date(value)
    except ValueError:
        return None
    return f"{value}T00:00:00+09:00"


def _timestamp_from_fields(
    row: Mapping[str, Any], contract: DatasetContract, fields: Sequence[str]
) -> str | None:
    if not fields:
        return None
    first = _pick(row, contract, fields[0])
    if not isinstance(first, str) or not first:
        return None
    if "T" in first or " " in first:
        return first
    if len(fields) < 2:
        return None
    second = _pick(row, contract, fields[1])
    if not isinstance(second, str) or not second or not _DATE_RE.fullmatch(first):
        return None
    try:
        _parse_date(first)
    except ValueError:
        return None
    clock = second
    if re.fullmatch(r"\d{2}:\d{2}", clock):
        clock += ":00"
    if not re.fullmatch(r"\d{2}:\d{2}:\d{2}(?:\.\d+)?", clock):
        return None
    return f"{first}T{clock}+09:00"


def event_time_for(row: Mapping[str, Any], dataset_id: str) -> str | None:
    contract = contract_for(dataset_id)
    if contract.event_time_policy == "session_close":
        value = _pick(row, contract, contract.event_time_fields[0])
        if isinstance(value, str) and _DATE_RE.fullmatch(value):
            try:
                return session_close_jst(value, session=contract.session)
            except ValueError:
                return None
        return None
    if contract.event_time_policy == "explicit_timestamp_field":
        instant = _timestamp_from_fields(row, contract, contract.event_time_fields)
        if instant is not None:
            return instant
        value = _pick(row, contract, contract.event_time_fields[0])
        return _date_start(value)
    if contract.event_time_policy == "observation_date":
        value = _pick(row, contract, contract.event_time_fields[0])
        return _date_start(value)
    return None


def _disclosure_available_at(
    row: Mapping[str, Any], contract: DatasetContract
) -> str | None:
    field = contract.availability_field
    if not field:
        return None
    fields = tuple(field.split("+"))
    instant = _timestamp_from_fields(row, contract, fields)
    if instant is not None:
        return instant
    value = _pick(row, contract, fields[0])
    if not isinstance(value, str) or not _DATE_RE.fullmatch(value):
        return None
    try:
        day = _parse_date(value) + timedelta(days=1)
    except ValueError:
        return None
    return day.strftime("%Y-%m-%dT00:00:00+09:00")


def _known_lag_available_at(
    row: Mapping[str, Any], contract: DatasetContract
) -> str | None:
    lag = contract.known_publication_lag
    event = event_time_for(row, contract.dataset_id)
    if lag is None or event is None:
        return None
    match = _LAG_RE.fullmatch(lag)
    if not match:
        return None
    try:
        dt = datetime.fromisoformat(event)
    except ValueError:
        return None
    dt += timedelta(days=int(match.group(1) or 0), hours=int(match.group(2) or 0))
    return dt.isoformat(timespec="seconds")


def _source_capability_for(dataset_id: str) -> SourceCapabilityContract | None:
    try:
        return source_capability_contract_for(dataset_id)
    except KeyError:
        return None


def _event_or_session_date(
    row: Mapping[str, Any], contract: DatasetContract
) -> str | None:
    field = None
    if contract.event_time_fields:
        field = contract.event_time_fields[0]
    elif contract.availability_field:
        field = contract.availability_field.split("+")[0]
    if not field:
        return None
    value = _pick(row, contract, field)
    if not isinstance(value, str) or not _DATE_RE.fullmatch(value):
        return None
    try:
        _parse_date(value)
    except ValueError:
        return None
    return value


def _before_official_availability(
    row: Mapping[str, Any], dataset_id: str, contract: DatasetContract
) -> bool:
    capability = _source_capability_for(dataset_id)
    if capability is None:
        return False
    day = _event_or_session_date(row, contract)
    if day is None:
        return False
    return day < capability.earliest_official_availability


def available_at_for(
    row: Mapping[str, Any], dataset_id: str, ingested_at: str
) -> str:
    """Compute contract-selected availability, always failing safe to ingest.

    ingest_time_conservative keeps ingested_at when the historical publication
    instant is unknown (master Date is observation day, not publication).
    SourceCapabilityContract is consulted so a pre-official event/session date
    is not rewritten into a Date-derived available_at that would make those
    rows look PIT-eligible. Fail-safe remains ingested_at. PIT query clamp
    (not this function) is the membership gate for as_of before official start.
    """
    contract = contract_for(dataset_id)
    policy = contract.available_at_policy
    # Consult official start even when the timestamp stays ingest-time.
    pre_official = _before_official_availability(row, dataset_id, contract)
    if pre_official and policy in (
        "ingest_time_conservative",
        "calendar_prepublished",
    ):
        # Do not mint Date-derived PIT eligibility for pre-official rows.
        # Tip-only AM session_close is left to its policy so this function
        # does not invent historical AM/earnings eligibility.
        return ingested_at
    if policy == "session_close":
        value = _pick(row, contract, contract.availability_field or "Date")
        if isinstance(value, str) and _DATE_RE.fullmatch(value):
            try:
                return session_close_jst(value, session=contract.session)
            except ValueError:
                pass
    elif policy == "explicit_timestamp_field":
        field = contract.availability_field
        if field:
            instant = _timestamp_from_fields(row, contract, tuple(field.split("+")))
            if instant is not None:
                return instant
    elif policy == "explicit_disclosure_date":
        instant = _disclosure_available_at(row, contract)
        if instant is not None:
            return instant
    elif policy == "known_publication_lag":
        instant = _known_lag_available_at(row, contract)
        if instant is not None:
            return instant
    # calendar_prepublished and ingest_time_conservative both deliberately
    # retain ingest time when the historical publication instant is unknown.
    return ingested_at


__all__ = [
    "available_at_for",
    "canonical_json",
    "event_time_for",
    "natural_key",
    "session_close_jst",
    "sha256_fallback",
]
